diff --git a/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp b/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
--- a/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
+++ b/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
@@ -1623,9 +1623,12 @@
 }

 OpFoldResult GatherOp::fold(FoldAdaptor adaptor) {
+  auto resultType = cast<TensorType>(
+      prime_ir::maybeConvertPrimeIRToBuiltinType(
+          cast<ShapedType>(getResult().getType())));
   if (OpFoldResult reshapedSource = reshapeConstantSource(
-          llvm::dyn_cast_if_present<DenseElementsAttr>(adaptor.getSource()),
-          getResult().getType()))
+          dyn_cast_if_present<DenseElementsAttr>(adaptor.getSource()),
+          resultType))
     return reshapedSource;
   return {};
 }
@@ -1864,9 +1867,12 @@
 }

 OpFoldResult ReshapeOp::fold(FoldAdaptor adaptor) {
+  auto resultType = cast<TensorType>(
+      prime_ir::maybeConvertPrimeIRToBuiltinType(
+          cast<ShapedType>(getResult().getType())));
   if (OpFoldResult reshapedSource = reshapeConstantSource(
-          llvm::dyn_cast_if_present<DenseElementsAttr>(adaptor.getSource()),
-          getResult().getType()))
+          dyn_cast_if_present<DenseElementsAttr>(adaptor.getSource()),
+          resultType))
     return reshapedSource;

   // If the producer of operand 'source' is another 'tensor.reshape' op, use the
@@ -2773,9 +2779,12 @@
 }

 OpFoldResult ExtractSliceOp::fold(FoldAdaptor adaptor) {
+  auto resultType = cast<TensorType>(
+      prime_ir::maybeConvertPrimeIRToBuiltinType(
+          cast<ShapedType>(getResult().getType())));
   if (OpFoldResult reshapedSource = reshapeConstantSource(
-          llvm::dyn_cast_if_present<SplatElementsAttr>(adaptor.getSource()),
-          getResult().getType()))
+          dyn_cast_if_present<SplatElementsAttr>(adaptor.getSource()),
+          resultType))
     return reshapedSource;
   if (getSourceType() == getType() &&
       succeeded(foldIdentityOffsetSizeAndStrideOpInterface(*this, getType())))
