diff --git mlir/lib/Dialect/Bufferization/IR/BufferizationOps.cpp mlir/lib/Dialect/Bufferization/IR/BufferizationOps.cpp
index 9bd87d66c7d3..0d0b41f51926 100644
--- mlir/lib/Dialect/Bufferization/IR/BufferizationOps.cpp
+++ mlir/lib/Dialect/Bufferization/IR/BufferizationOps.cpp
@@ -835,7 +835,7 @@ struct LoadOfToBuffer : public OpRewritePattern<memref::LoadOp> {
   LogicalResult matchAndRewrite(memref::LoadOp load,
                                 PatternRewriter &rewriter) const override {
     auto toBuffer = load.getMemref().getDefiningOp<ToBufferOp>();
-    if (!toBuffer)
+    if (!toBuffer || !toBuffer.getReadOnly())
       return failure();

     rewriter.replaceOpWithNewOp<tensor::ExtractOp>(load, toBuffer.getTensor(),
