diff --git mlir/lib/Dialect/Tensor/IR/TensorOps.cpp mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
index 22a25fd1a5af..7dad917d5056 100644
--- mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
+++ mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
@@ -41,6 +41,9 @@
 #include <optional>
 #include <vector>
 
+#include "zkir/Dialect/ModArith/IR/ModArithOps.h"
+#include "zkir/IR/Attributes.h"
+
 using namespace mlir;
 using namespace mlir::tensor;
 
@@ -53,6 +56,9 @@ using llvm::mod;
 Operation *TensorDialect::materializeConstant(OpBuilder &builder,
                                               Attribute value, Type type,
                                               Location loc) {
+  if (auto op =
+          zkir::mod_arith::ConstantOp::materialize(builder, value, type, loc))
+    return op;
   if (auto op = arith::ConstantOp::materialize(builder, value, type, loc))
     return op;
   if (complex::ConstantOp::isBuildableWith(value, type))
@@ -1465,8 +1471,12 @@ void FromElementsOp::build(OpBuilder &builder, OperationState &result,
 }
 
 OpFoldResult FromElementsOp::fold(FoldAdaptor adaptor) {
-  if (!llvm::is_contained(adaptor.getElements(), nullptr))
+  if (!llvm::is_contained(adaptor.getElements(), nullptr)) {
+    if (zkir::ZkirDenseElementsAttr::isValidType(getType())) {
+      return zkir::ZkirDenseElementsAttr::get(getType(), adaptor.getElements());
+    }
     return DenseElementsAttr::get(getType(), adaptor.getElements());
+  }
   return {};
 }
 
diff --git utils/bazel/llvm-project-overlay/mlir/BUILD.bazel utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
index 1bf8488d475b..9286fc7a06c1 100644
--- utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
+++ utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
@@ -7135,6 +7135,8 @@ cc_library(
         ":ValueBoundsOpInterface",
         ":ViewLikeInterface",
         "//llvm:Support",
+        "@zkir//zkir/Dialect/ModArith/IR:ModArith",
+        "@zkir//zkir/IR:AttributesHdrs",
     ],
 )
 
