diff --git a/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp b/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
--- a/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
+++ b/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
@@ -37,6 +37,10 @@
 #include "llvm/ADT/StringRef.h"
 #include "llvm/Support/Casting.h"
 #include "llvm/Support/MathExtras.h"
+
+#include "prime_ir/Dialect/Field/IR/FieldOps.h"
+#include "prime_ir/Dialect/ModArith/IR/ModArithOps.h"
+#include "prime_ir/IR/Attributes.h"
 #include <optional>

 using namespace mlir;
@@ -47,6 +51,11 @@ using namespace mlir::tensor;
 Operation *TensorDialect::materializeConstant(OpBuilder &builder,
                                               Attribute value, Type type,
                                               Location loc) {
+  if (auto op = prime_ir::field::ConstantOp::materialize(builder, value, type, loc))
+    return op;
+  if (auto op =
+          prime_ir::mod_arith::ConstantOp::materialize(builder, value, type, loc))
+    return op;
   if (auto op = arith::ConstantOp::materialize(builder, value, type, loc))
     return op;
   if (complex::ConstantOp::isBuildableWith(value, type))
@@ -1457,8 +1466,10 @@ void FromElementsOp::build(OpBuilder &builder, OperationState &result,
 }

 OpFoldResult FromElementsOp::fold(FoldAdaptor adaptor) {
-  if (!llvm::is_contained(adaptor.getElements(), nullptr))
-    return DenseElementsAttr::get(getType(), adaptor.getElements());
+  if (!llvm::is_contained(adaptor.getElements(), nullptr)) {
+    ShapedType type = prime_ir::maybeConvertPrimeIRToBuiltinType(getType());
+    return DenseElementsAttr::get(type, adaptor.getElements());
+  }
   return {};
 }

@@ -4059,7 +4070,9 @@ OpFoldResult SplatOp::fold(FoldAdaptor adaptor) {

   // SplatElementsAttr::get treats single value for second arg as being a
   // splat.
-  return SplatElementsAttr::get(getType(), {constOperand});
+  ShapedType destType =
+      cast<ShapedType>(prime_ir::maybeConvertPrimeIRToBuiltinType(getType()));
+  return SplatElementsAttr::get(destType, {constOperand});
 }

 //===----------------------------------------------------------------------===//
diff --git a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
--- a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
@@ -7682,6 +7682,9 @@ cc_library(
         ":ValueBoundsOpInterface",
         ":ViewLikeInterface",
         "//llvm:Support",
+        "@prime_ir//prime_ir/Dialect/Field/IR:FieldHdrs",
+        "@prime_ir//prime_ir/Dialect/ModArith/IR:ModArithHdrs",
+        "@prime_ir//prime_ir/IR:Attributes",
     ],
 )

