diff --git a/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp b/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
index 22a25fd1a5af..1832036990e4 100644
--- a/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
+++ b/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
@@ -41,6 +41,10 @@
 #include <optional>
 #include <vector>
 
+#include "prime_ir/Dialect/Field/IR/FieldOps.h"
+#include "prime_ir/Dialect/ModArith/IR/ModArithOps.h"
+#include "prime_ir/IR/Attributes.h"
+
 using namespace mlir;
 using namespace mlir::tensor;
 
@@ -53,6 +57,11 @@ using llvm::mod;
 Operation *TensorDialect::materializeConstant(OpBuilder &builder,
                                               Attribute value, Type type,
                                               Location loc) {
+  if (auto op = prime_ir::field::ConstantOp::materialize(builder, value, type, loc))
+    return op;
+  if (auto op =
+          prime_ir::mod_arith::ConstantOp::materialize(builder, value, type, loc))
+    return op;
   if (auto op = arith::ConstantOp::materialize(builder, value, type, loc))
     return op;
   if (complex::ConstantOp::isBuildableWith(value, type))
@@ -1465,8 +1474,10 @@ void FromElementsOp::build(OpBuilder &builder, OperationState &result,
 }
 
 OpFoldResult FromElementsOp::fold(FoldAdaptor adaptor) {
-  if (!llvm::is_contained(adaptor.getElements(), nullptr))
-    return DenseElementsAttr::get(getType(), adaptor.getElements());
+  if (!llvm::is_contained(adaptor.getElements(), nullptr)) {
+    ShapedType type = prime_ir::maybeConvertPrimeIRToBuiltinType(getType());
+    return DenseElementsAttr::get(type, adaptor.getElements());
+  }
   return {};
 }
 
@@ -4052,7 +4063,9 @@ OpFoldResult SplatOp::fold(FoldAdaptor adaptor) {
 
   // SplatElementsAttr::get treats single value for second arg as being a
   // splat.
-  return SplatElementsAttr::get(getType(), {constOperand});
+  ShapedType destType =
+      cast<ShapedType>(prime_ir::maybeConvertPrimeIRToBuiltinType(getType()));
+  return SplatElementsAttr::get(destType, {constOperand});
 }
 
 //===----------------------------------------------------------------------===//
diff --git a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
index 7829b11da096..e86158d51727 100644
--- a/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
+++ b/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
@@ -7134,6 +7134,9 @@ cc_library(
         ":ValueBoundsOpInterface",
         ":ViewLikeInterface",
         "//llvm:Support",
+        "@prime_ir//prime_ir/Dialect/Field/IR:FieldHdrs",
+        "@prime_ir//prime_ir/Dialect/ModArith/IR:ModArithHdrs",
+        "@prime_ir//prime_ir/IR:Attributes",
     ],
 )
 
