diff --git mlir/lib/Dialect/Tensor/IR/TensorOps.cpp mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
index 22a25fd1a5af..bbfab789d58a 100644
--- mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
+++ mlir/lib/Dialect/Tensor/IR/TensorOps.cpp
@@ -41,6 +41,8 @@
 #include <optional>
 #include <vector>

+#include "zkir/Dialect/ModArith/IR/ModArithTypes.h"
+
 using namespace mlir;
 using namespace mlir::tensor;

@@ -1465,8 +1467,16 @@ void FromElementsOp::build(OpBuilder &builder, OperationState &result,
 }

 OpFoldResult FromElementsOp::fold(FoldAdaptor adaptor) {
-  if (!llvm::is_contained(adaptor.getElements(), nullptr))
+  if (!llvm::is_contained(adaptor.getElements(), nullptr)) {
+    if (auto modArithType = dyn_cast<zkir::mod_arith::ModArithType>(
+            getType().getElementType())) {
+      auto st = cast<ShapedType>(getType());
+      return DenseElementsAttr::get(
+          st.cloneWith(std::nullopt, modArithType.getStorageType()),
+          adaptor.getElements());
+    }
     return DenseElementsAttr::get(getType(), adaptor.getElements());
+  }
   return {};
 }

diff --git utils/bazel/llvm-project-overlay/mlir/BUILD.bazel utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
index cc266c2fe3a7..9ef338d525f1 100644
--- utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
+++ utils/bazel/llvm-project-overlay/mlir/BUILD.bazel
@@ -7134,5 +7134,6 @@ cc_library(
         ":ValueBoundsOpInterface",
         ":ViewLikeInterface",
         "//llvm:Support",
+        "@zkir//zkir/Dialect/ModArith/IR:ModArith",
     ],
 )
